### PR TITLE
Adds additional loss functions

### PR DESCRIPTION
**Pull request by georgeyiasemis**
Adds:
* High frequency error norm (hfen) as loss and metric
* Signal-to-noise ratio (snr) loss
* SSIM 3d loss
* Tests

### CODE REVIEW

File: direct/functionals/__init__.py [hunk 1]
```diff
@@ -1,9 +1,12 @@
...
+"""direct.nn.functionals module."""
+
 from direct.functionals.challenges import *
 from direct.functionals.grad import *
+from direct.functionals.hfen import *
 from direct.functionals.nmae import NMAELoss
 from direct.functionals.nmse import *
 from direct.functionals.psnr import *
+from direct.functionals.snr import *
 from direct.functionals.ssim import *
```
Comment (jonasteuwen) [inline direct/functionals/__init__.py:12]: Typically you would define an `__all__` if you do it this way.

File: direct/functionals/ssim.py [hunk 2]
```diff
@@ -38,15 +42,15 @@ def __init__(self, win_size=7, k1=0.01, k2=0.03):
         NP = win_size**2
         self.cov_norm = NP / (NP - 1)
 
-    def forward(self, X, Y, data_range):
+    def forward(self, input_data: torch.Tensor, target_data: torch.Tensor, data_range: torch.Tensor) -> torch.Tensor:
         data_range = data_range[:, None, None, None]
```
Comment (jonasteuwen) [inline direct/functionals/ssim.py:46]: Are you assuming this will always be a float32? I believe the difference between torch.tensor and torch.Tensor is that the latter is fp32.

File: direct/nn/mri_models.py [hunk 3]
```diff
@@ -423,6 +600,18 @@ def grad_l2_loss(
...
+                    curr_loss.multiplier, (snr_loss if loss_fn == "snr_loss" else psnr_loss)
+                )
+            elif loss_fn == "hfen_l1_loss":
+                loss_dict[loss_fn] = multiply_function(curr_loss.multiplier, hfen_l1_loss)
+            elif loss_fn == "hfen_l2_loss":
+                loss_dict[loss_fn] = multiply_function(curr_loss.multiplier, hfen_l2_loss)
+            elif loss_fn == "hfen_l1_norm_loss":
+                loss_dict[loss_fn] = multiply_function(curr_loss.multiplier, hfen_l1_norm_loss)
+            elif loss_fn == "hfen_l2_norm_loss":
+                loss_dict[loss_fn] = multiply_function(curr_loss.multiplier, hfen_l2_norm_loss)
```
Comment (jonasteuwen) [inline direct/nn/mri_models.py:614]: Perhaps it's better tp turn this into a NamedTuple, or some sort, subtyping it also from string. Then you can loss_fn == that string, but also do loss_fn calling directly as a function.

**Review by jonasteuwen — COMMENTED** [review]
Looks good, but some minor comments.

File: direct/functionals/ssim.py (hunk 2 above)
Comment (georgeyiasemis) [reply direct/functionals/ssim.py:46]: I am not sure if this is the case but the whole software has this typing concerning Tensors, no?

Comment (georgeyiasemis) [reply direct/functionals/ssim.py:46]: Actually this is the behaviour

```
>>> x = torch.arange(10, dtype=torch.float16)
>>> isinstance(x, torch.Tensor)
True
>>> x = torch.arange(10, dtype=torch.float32)
>>> isinstance(x, torch.Tensor)
True
```

File: direct/functionals/__init__.py (hunk 1 above)
Comment (georgeyiasemis) [reply direct/functionals/__init__.py:37]: Done!

File: direct/nn/mri_models.py (hunk 3 above)
Comment (georgeyiasemis) [reply direct/nn/mri_models.py]: Done!

**Review by jonasteuwen — APPROVED** [review]
LGTM
